### PR TITLE
Block deployments when `strict` preflights fail

### PR DESCRIPTION
**Pull request by pavansokkenagaraj**
#### What type of PR is this?
**type::feature**

#### What this PR does / why we need it:
[SC-24474](https://app.shortcut.com/replicated/story/24474/vendor-ability-to-prevent-an-end-user-from-ignoring-failing-app-preflights)

- [x] troubleshoot changes PR : https://github.com/replicatedhq/troubleshoot/pull/539 
- [x] troubleshoot docs PR: https://github.com/replicatedhq/troubleshoot.sh/pull/364
- [ ] kots docs PR: https://github.com/replicatedhq/replicated-docs/pull/212
- [x] testim test suite merge to master: https://app.testim.io/#/project/wpYAooUimFDgQxY73r17/branch/pa1-strict-test/automate/suites/gaC28U0qJCMGH9H5
- [ ] new test app PR: https://github.com/replicatedhq/kots-test-apps/pull/20
- [x] testim test suite/build-job passing? https://github.com/replicatedhq/kots/actions/runs/1966600210

#### Special notes for your reviewer:
Code changes:
 - persist rendered preflight_spec value in `app_version`  instead of non-rendered values
 - if there are strict preflights, override query param `SkipPreflights=true` and run preflights in all the flows (online/airgap/upstream-update/upload/automatic-update/config-changes/)
 - during deployment, if downstream_version status is `pending_preflight`, poll every 2 seconds for status to change to `pending` and validate the preflight results to not have failing strict preflights and deploy the version
 - add new response boolean param `HasFailingStrictPreflights` to endpoints  `/apps` `/app` `/versions` which can be used to disable deployment button on UI and mandate running preflights and making sure strict preflights are not failing
 - Use `HasFailingStrictPreflights` and disable `deploy` button on UI

#### Scenarios to test/tested
- [x] UI - airgap install 
- [x] UI -online install
- [x] UI - config change which will enable/disable strict preflights
- [x] CLI - upstream-upgrade
- [x] CLI - config change which will enable/disable strict preflights
- [x] CLI - upload a new version
- [x] CRON - automatic upgrades
- [x] Minimal RBAC installation

#### Does this PR introduce a user-facing change?
**YES**

```release-note
- ability to provide a strict preflight spec which should run and not fail before deploying an app.
- override flag `--skip-preflights=true` and run preflights when there are strict preflights present.
- disable `deploy button` when there is a strict preflights spec and if preflight is still running or strict preflight has failed.
- when flag `--deploy` is provided or deployment is triggered, if preflights are still running and has a strict preflight, poll until the preflights are finished running and analyze the preflight result for any failed strict preflights. If a strict preflight has failed, deployment is not continued and an appropriate error message is returned. 
- if preflights are running and deployment is triggered, deployment will poll for preflights execution to finish with a timeout of 15 minutes.
```

#### Does this PR require documentation?
**YES**
https://github.com/replicatedhq/replicated-docs/pull/207

### CODE REVIEW

**Comment by pavansokkenagaraj** [comment]
# kots

## kots upstream upgrade
| CMD |has strict preflights |--skip-preflight | --deploy | run preflights | wait for preflight results | strict preflight fails | Block Deploy |   DEPLOY | 
| - | - | - | - | - | - | - | - | - |  
| upstream upgrade | NO | true | false | false | NO | NO | NO | NO |    
| upstream upgrade | NO | true | true | false | NO | NO | NO | YES |
| upstream upgrade | NO | false | true | true | NO | NO | NO | YES |
| upstream upgrade | NO | false | false | true | NO | NO | NO | NO |
| - | - | - | - | - | - | - | - | - |  


| CMD |has strict preflights |--skip-preflight | --deploy | run preflights | wait for preflight results | strict preflight fails | Block Deploy |   DEPLOY | 
| - | - | - | - | - | - | - | - | - |  
| upstream upgrade | YES | true | false | OVERRIDE [true] | NO | NO | NO | NO |    
| upstream upgrade | YES | true | true | OVERRIDE [true] | YES | NO | NO | YES |
| upstream upgrade | YES | true | true | OVERRIDE [true] | YES | YES | YES | NO |
| upstream upgrade | YES | false | true | true | YES | NO | NO | YES |
| upstream upgrade | YES | false | true | true | YES | YES | YES | NO |
| upstream upgrade | YES | false | false | true | NO | NO | NO | NO |
| - | - | - | - | - | - | - | - | - |  


## kots upload
| CMD |has strict preflights |--skip-preflight | --deploy | run preflights | wait for preflight results | strict preflight fails | Block Deploy |   DEPLOY | 
| - | - | - | - | - | - | - | - | - |  
| upload | NO | true | false | false | NO | NO | NO | NO |    
| upload | NO | true | true | false | NO | NO | NO | YES |
| upload | NO | false | true | true | NO | NO | NO | YES |
| upload | NO | false | false | true | NO | NO | NO | NO |
| - | - | - | - | - | - | - | - | - |  


| CMD |has strict preflights |--skip-preflight | --deploy | run preflights | wait for preflight results | strict preflight fails | Block Deploy |   DEPLOY | 
| - | - | - | - | - | - | - | - | - |  
| upload | YES | true | false | OVERRIDE [true] | NO | NO | NO | NO |    
| upload | YES | true | true | OVERRIDE [true] | YES | NO | NO | YES |
| upload | YES | true | true | OVERRIDE [true] | YES | YES | YES | NO |
| upload | YES | false | true | true | YES | NO | NO | YES |
| upload | YES | false | true | true | YES | YES | YES | NO |
| upload | YES | false | false | true | NO | NO | NO | NO |
| - | - | - | - | - | - | - | - | - |  

## kots set config
| CMD |has strict preflights |--skip-preflight | --deploy | run preflights | wait for preflight results | strict preflight fails | Block Deploy |   DEPLOY | 
| - | - | - | - | - | - | - | - | - |  
| set config | NO | true | false | false | NO | NO | NO | NO |    
| set config | NO | true | true | false | NO | NO | NO | YES |
| set config | NO | false | true | true | NO | NO | NO | YES |
| set config | NO | false | false | true | NO | NO | NO | NO |
| - | - | - | - | - | - | - | - | - |  


| CMD |has strict preflights |--skip-preflight | --deploy | run preflights | wait for preflight results | strict preflight fails | Block Deploy |   DEPLOY | 
| - | - | - | - | - | - | - | - | - |  
| set config | YES | true | false | OVERRIDE [true] | NO | NO | NO | NO |    
| set config | YES | true | true | OVERRIDE [true] | YES | NO | NO | YES |
| set config | YES | true | true | OVERRIDE [true] | YES | YES | YES | NO |
| set config | YES | false | true | true | YES | NO | NO | YES |
| set config | YES | false | true | true | YES | YES | YES | NO |
| set config | YES | false | false | true | NO | NO | NO | NO |
| - | - | - | - | - | - | - | - | - |  

## Configure automatic updates [GUI]
- calling it cron for easier understanding
- automatic updates run preflights and don't wait for preflights to finish, but the deploy will happen
- So, `--skip-preflight` will be alway `false`

| CMD |has strict preflights |--skip-preflight | --deploy | run preflights | wait for preflight results | strict preflight fails | Block Deploy |   DEPLOY | 
| - | - | - | - | - | - | - | - | - |  
| cron | NO | false | true | true | NO | NO | NO | YES |
| - | - | - | - | - | - | - | - | - |  

| CMD |has strict preflights | --skip-preflight | --deploy | run preflights | wait for preflight results | strict preflight fails | Block Deploy |   DEPLOY | 
| - | - | - | - | - | - | - | - | - |  
| cron | YES | false | true | true | YES | NO | NO | YES |
| cron | YES | false | true | true | YES | YES | YES | NO |
| - | - | - | - | - | - | - | - | - |

**Comment by pavansokkenagaraj** [comment]
**CLI ouput**
```terminal
➜  ~ which kkots
kkots: aliased to /Users/sokke/go/src/github.com/replicatedhq/kots/bin/kots-local
➜  ~ kkots  upload --namespace pavansokkenagaraj --slug pa1-test ./pa1-test --deploy
  • Uploading local application to Admin Console ✗
  • Unable to deploy as preflight check's strict analyzer has failed
Error: Unexpected response from the API: 500

➜  ~ kkots upstream upgrade -n pavansokkenagaraj  pa1-test --deploy --skip-preflights=true
  • Checking for application updates ✗
  • Unable to deploy as preflight check's strict analyzer has failed
  • Unexpected response from the API: 500
➜  ~
```

**Comment by markpundsack** [comment]
@pavansokkenagaraj Can you elaborate on why "persist rendered preflight_spec value in app_version instead of non-rendered values" is necessary?

**Comment by pavansokkenagaraj** [comment]
> @pavansokkenagaraj Can you elaborate on why "persist rendered preflight_spec value in app_version instead of non-rendered values" is necessary?

@markpundsack as discussed in [SC-24474-comments](https://app.shortcut.com/replicated/story/24474/vendor-ability-to-prevent-an-end-user-from-ignoring-failing-app-preflights#activity-43988)

`app_version` table's `preflight_spec` column holds non-rendered preflight spec.

If preflight specs are rendered using config values, we will not be able to check if a preflight_spec has strict: true/false
So, we would need to persist the rendered preflight spec to get the right preflight_yaml

**Comment by markpundsack** [comment]
@pavansokkenagaraj Thanks, that makes sense. I'm trying to figure out why the preflights were stored unrendered in the first place, and more specifically, is there any chance we'll regret storing them rendered? Like does something depend on being able to re-render them after making a config change in a subsequent version and now the original non-rendered version is lost? Anyway, not a blocker; I'm just curious (and cautious).

**Comment by pavansokkenagaraj** [comment]
@markpundsack 
> why the preflights were stored unrendered in the first place

Even, I don't know the reason why we were storing unrendered preflight_spec

> is there any chance we'll regret storing them rendered? 

I went through the code and usage and couldn't find anything.
There is no dependency the stored spec is not used anywhere, so there would be no issues.
During migrations, the spec would be re-loaded from the directory files

File: pkg/airgap/airgap.go [hunk 1]
```diff
@@ -265,6 +266,21 @@ func CreateAppFromAirgap(opts CreateAirgapAppOpts) (finalError error) {
...
+	if err != nil {
+		return errors.Wrap(err, "failed to check if app preflight has strict analyzers")
+	}
+
+	if hasStrictPreflights && opts.SkipPreflights {
+		logger.Warnf("preflights will not be skipped, strict preflights are set to %t", hasStrictPreflights)
+	}
+
+	if !opts.SkipPreflights || hasStrictPreflights {
+		if err := preflight.Run(opts.PendingApp.ID, opts.PendingApp.Slug, newSequence, true, tmpRoot); err != nil {
```
Comment (sgalsaleh) [inline pkg/airgap/airgap.go:279]: hmmm, since this is moved here, let's remove lines 302-307 that sets the version as "pending preflight"? this step is already done in the preflight "Run" function, and we don't wanna end up with a race condition where it's overwritten afterwards.

File: pkg/kotsutil/kots.go [hunk 2]
```diff
@@ -162,6 +163,47 @@ func (k *KotsKinds) HasPreflights() bool {
 	return len(k.Preflight.Spec.Analyzers) > 0
 }
 
+func HasStrictPreflights(preflight *troubleshootv1beta2.Preflight) bool {
+	if preflight == nil {
+		return false
+	}
+	for _, a := range preflight.Spec.Analyzers {
+		if (a.ClusterVersion != nil && a.ClusterVersion.Strict.BoolOrDefaultFalse()) ||
```
Comment (sgalsaleh) [inline pkg/kotsutil/kots.go:171]: i still believe that this can be easily abstracted instead of checking for each individual analyzer

Comment (sgalsaleh) [reply pkg/kotsutil/kots.go:171]: which would be great if we add new analyzers later

File: pkg/online/online.go [hunk 3]
```diff
@@ -186,6 +187,21 @@ func CreateAppFromOnline(opts CreateOnlineAppOpts) (_ *kotsutil.KotsKinds, final
...
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to check if app preflight has strict analyzers")
+	}
+
+	if hasStrictPreflights && opts.SkipPreflights {
+		logger.Warnf("preflights will not be skipped, strict preflights are set to %t", hasStrictPreflights)
+	}
+
+	if !opts.SkipPreflights || hasStrictPreflights {
+		if err := preflight.Run(opts.PendingApp.ID, opts.PendingApp.Slug, newSequence, false, tmpRoot); err != nil {
```
Comment (sgalsaleh) [inline pkg/online/online.go:200]: since as earlier comment, let's remove lines 225-230?

File: pkg/store/kotsstore/version_store.go [hunk 4]
```diff
@@ -1007,3 +1014,54 @@ func (s *KOTSStore) GetCurrentUpdateCursor(appID string, channelID string) (stri
 
 	return updateCursor.String, versionLabel.String, nil
 }
+
+func (s *KOTSStore) HasStrictPreflights(appID string, sequence int64) (bool, error) {
+
```
Comment (sgalsaleh) [inline pkg/store/kotsstore/version_store.go:1019]: let's stick to the pattern of not having empty lines at the beginning of functions? :)

File: pkg/store/kotsstore/version_store.go [hunk 5]
```diff
@@ -1007,3 +1014,54 @@ func (s *KOTSStore) GetCurrentUpdateCursor(appID string, channelID string) (stri
...
+		}
+		return kotsutil.HasStrictPreflights(preflight), nil
+	}
+	// no preflight spec, so return false, nil
+	return false, nil
+}
+
+func (s *KOTSStore) renderPreflightSpec(appID string, appSlug string, sequence int64, isAirgap bool, kotsKinds *kotsutil.KotsKinds, renderer rendertypes.Renderer) (*troubleshootv1beta2.Preflight, error) {
+
+	if kotsKinds.Preflight != nil {
```
Comment (sgalsaleh) [inline pkg/store/kotsstore/version_store.go:1042]:
```suggestion
	if kotsKinds.HasPreflights() {
```

File: web/src/components/PreflightResultPage.jsx [hunk 6]
```diff
@@ -314,6 +320,16 @@ class PreflightResultPage extends Component {
 
     const preflightSkipped = preflightResultData?.skipped;
     const stopPolling = (preflightResultData?.result || preflightSkipped);
+    let blockDeployment = false;
+    if (this.props.fromLicenseFlow) {
+      for (const app of this.props.appsList) {
+        if (app.slug == slug) {
+          blockDeployment = app.downstreams[0]?.pendingVersions[0]?.hasFailingStrictPreflights;
```
Comment (sgalsaleh) [inline web/src/components/PreflightResultPage.jsx:327]: why not just return this as part of the preflight results response? you already have the version sequence there. adding the "list app" job isn't needed.

File: pkg/version/version.go [hunk 7]
```diff
@@ -199,3 +216,106 @@ func GetForwardedPortsFromAppSpec(appID string, sequence int64) ([]types.Forward
...
+	row := db.QueryRow(query, appID, sequence)
+	if err := row.Scan(
+		&status,
+	); err != nil {
+		return "", errors.Wrap(err, "failed to scan")
+	}
+	return storetypes.DownstreamVersionStatus(status.String), nil
+}
+
+func isBlockedDueToStricPreFlights(appID string, sequence int64) (bool, error) {
```
[the remark this answers is not in the document]
Comment (pavansokkenagaraj) [reply pkg/version/version.go]: `isBlockedDueToStrictPreFlights` is doing more than fetching stored data
if status is `pending_preflight`, it will poll for status every 2 seconds till it is changed to `pending`
and analyze the preflight result and validate if preflight has strict and preflight result are not failing for strict.

....
and after writing the above, i realize that I could reuse `func (s *KOTSStore) GetCurrentVersion` and do the same 🤦🏽

[the remark this answers is not in the document]
Comment (pavansokkenagaraj) [reply pkg/version/version.go]: I could have re-used existing store funcs, changing the code to re-use store funcs

File: web/src/components/PreflightResultPage.jsx (hunk 6 above)
Comment (pavansokkenagaraj) [reply web/src/components/PreflightResultPage.jsx:327]: `hasFailingStrictPreflights`  is calculated from `app_version.preflight_spec` and `app_downstream_version.preflght_result`.
Will it be okay to joint these table with appID and sequence?

Comment (pavansokkenagaraj) [reply web/src/components/PreflightResultPage.jsx]: added a new field `hasFailingStrictPreflights` to Preflight response and cleaned up the UI code.
Thanks for the suggestion

File: pkg/store/kotsstore/preflight_store.go [hunk 8]
```diff
@@ -61,16 +64,19 @@ func (s *KOTSStore) GetPreflightResults(appID string, sequence int64) (*prefligh
...
 		app_downstream_version.preflight_skipped,
 		app.slug as app_slug,
-		cluster.slug as cluster_slug
+		cluster.slug as cluster_slug,
+		app_version.preflight_spec 
 	FROM app_downstream_version
 		INNER JOIN app ON app_downstream_version.app_id = app.id
 		INNER JOIN cluster ON app_downstream_version.cluster_id = cluster.id
+		INNER JOIN app_version ON app_downstream_version.app_id = app_version.app_id 
+							AND app_downstream_version.parent_sequence = app_version.sequence
```
Comment (sgalsaleh) [inline pkg/store/kotsstore/preflight_store.go:73]: indentation here wrong. maybe just have on the same line of join?

File: pkg/kotsutil/kots.go (hunk 2 above)
Comment (sgalsaleh) [reply pkg/kotsutil/kots.go:172]: discussed in slack, to be updated

File: .github/workflows/build-test.yaml [hunk 9]
```diff
@@ -779,4 +779,85 @@ jobs:
...
+        if: startsWith(github.ref, 'refs/heads/')
+        id: get_testim_branch
+        run: |
+          BRANCH=${GITHUB_REF/refs\/heads\//}
+          if [ "$BRANCH" == "main" ]; then
+            BRANCH="master"
+          fi
+          echo ::set-output name=TESTIM_BRANCH::${BRANCH:-master}
+
+      - name: prepare strict-preflight-checks online install
```
Comment (sgalsaleh) [inline .github/workflows/build-test.yaml:830]: i don't see this test in testim. not on main nor on your branch.

File: .github/workflows/pr-test.yaml [hunk 10]
```diff
@@ -0,0 +1,305 @@
+name: pr-test
+
+# on:
```
Comment (sgalsaleh) [inline .github/workflows/pr-test.yaml:3]: if you're done, remove this file?

File: pkg/handlers/config.go [hunk 11]
```diff
@@ -591,7 +591,17 @@ func updateAppConfig(updateApp *apptypes.App, sequence int64, configGroups []kot
 		return updateAppConfigResponse, err
 	}
 
-	if !skipPreflights {
+	hasStrictPreflights, err := store.GetStore().HasStrictPreflights(updateApp.ID, sequence)
+	if err != nil {
+		updateAppConfigResponse.Error = errors.Cause(err).Error()
```
Comment (sgalsaleh) [inline pkg/handlers/config.go:596]: let's just return a hardcoded message here? we don't know what the stacktrace will have, and it's not a best practice to return anything from the stacktrace to the user.
```suggestion
		updateAppConfigResponse.Error = "failed to check if version has strict preflights"
```

File: pkg/kotsutil/kots.go [hunk 12]
```diff
@@ -163,6 +165,56 @@ func (k *KotsKinds) HasPreflights() bool {
...
+
+	analyzersMap := []map[string]interface{}{}
+	err = json.Unmarshal(marshalledAnalyzers, &analyzersMap) // Unmarshall again so we can loop over non nil analyzers
+	if err != nil {
+		return false, errors.Wrap(err, "error while un-marshalling marshalledAnalyzers")
+	}
+
+	// analyzerMap will ignore empty Analyzers and loop around Analyzer with data
+	for _, analyzers := range analyzersMap { // for each analyzers: map["clusterVersion": map[string]interface{} ["exclude": "", "strict": "true", "outcomes": nil]
+		for _, analyzer := range analyzers { // for each analyzeMeta: map[string]interface{} ["exclude": "", "strict": "true", "outcomes": nil]
```
Comment (sgalsaleh) [inline pkg/kotsutil/kots.go:186]: how is the example here for analyzeMeta any different than the one line above?

File: pkg/kotsutil/kots.go [hunk 13]
```diff
@@ -163,6 +165,56 @@ func (k *KotsKinds) HasPreflights() bool {
...
+			marshalledAnalyzer, err := json.Marshal(analyzer)
+			if err != nil {
+				return false, errors.Wrap(err, "error while marshalling analyzer")
+			}
+			// return Analyzer.Strict which can be extraceted from AnalyzeMeta
```
Comment (sgalsaleh) [inline pkg/kotsutil/kots.go:191]:
```suggestion
			// return Analyzer.Strict which can be extracted from AnalyzeMeta
```

File: pkg/kotsutil/kots.go [hunk 14]
```diff
@@ -163,6 +165,56 @@ func (k *KotsKinds) HasPreflights() bool {
...
 }
 
+func HasStrictPreflights(preflight *troubleshootv1beta2.Preflight) (bool, error) {
+	if preflight == nil {
+		return false, nil
+	}
+
+	marshalledAnalyzers, err := json.Marshal(preflight.Spec.Analyzers) // marshall and remove nil Analyzers eg result: "[{\"clusterVersion\":{\"exclude\":\"\",\"strict\":\"false\",\"outcomes\":null}}]"
+	if err != nil {
+		return false, errors.Wrap(err, "error while marshalling preflight.Spec.Analyzers")
```
Comment (sgalsaleh) [inline pkg/kotsutil/kots.go:175]:
```suggestion
		return false, errors.Wrap(err, "failed to marshal analyzers")
```

File: pkg/kotsutil/kots.go [hunk 15]
```diff
@@ -163,6 +165,56 @@ func (k *KotsKinds) HasPreflights() bool {
...
+	}
+
+	analyzersMap := []map[string]interface{}{}
+	err = json.Unmarshal(marshalledAnalyzers, &analyzersMap) // Unmarshall again so we can loop over non nil analyzers
+	if err != nil {
+		return false, errors.Wrap(err, "error while un-marshalling marshalledAnalyzers")
```
Comment (sgalsaleh) [inline pkg/kotsutil/kots.go:181]:
```suggestion
		return false, errors.Wrap(err, "failed to unmarshal analyzers")
```

File: pkg/kotsutil/kots.go [hunk 16]
```diff
@@ -163,6 +165,56 @@ func (k *KotsKinds) HasPreflights() bool {
...
+	}
+
+	// analyzerMap will ignore empty Analyzers and loop around Analyzer with data
+	for _, analyzers := range analyzersMap { // for each analyzers: map["clusterVersion": map[string]interface{} ["exclude": "", "strict": "true", "outcomes": nil]
```
Comment (sgalsaleh) [inline pkg/kotsutil/kots.go:185]: about the example in the comment, this go format is hard to parse, can you just post the example in json?
```suggestion
	for _, analyzers := range analyzersMap { // for each analyzer: map["clusterVersion": map[string]interface{} ["exclude": "", "strict": "true", "outcomes": nil]
```

File: pkg/kotsutil/kots.go (hunk 12 above)
Comment (sgalsaleh) [inline pkg/kotsutil/kots.go:186]: same here

**Review by sgalsaleh — APPROVED** [review]
